vizro-ai/src/vizro_ai/dashboard/utils.py

```diff
@@ -61,3 +63,30 @@ def _register_data(all_df_metadata: AllDfMetadata) -> vm.Dashboard:
 
     for name, metadata in all_df_metadata.all_df_metadata.items():
         data_manager[name] = metadata.df
+
+
+def _dashboard_code(
+    dashboard: vm.Dashboard, extra_callable_defs: Optional[Set[str]] = None, extra_imports: Optional[Set[str]] = None
+) -> str:
+    """Get the code for the dashboard."""
+    return dashboard._to_python(extra_callable_defs=extra_callable_defs, extra_imports=extra_imports)
+
+
+def _process_to_set(list_of_list_of_dict):
+    custom_func = set()
+    imports = set()
+    for list_of_dict in list_of_list_of_dict:
+        for d in list_of_dict:
+            for value in d.values():
+                # Extract imports
+                import_match = re.match(r"((?:from|import).*?)\n\n", value, re.DOTALL)
+                if import_match:
+                    imports.add(import_match.group(1))
+
+                # Remove leading imports and the last line (any chart creation)
+                value = re.sub(r"^(?:from|import).*?\n\n", "", value, flags=re.DOTALL)  # noqa: PLW2901
+                value = re.sub(r"\n\nfig = .*?\(data_frame=df\)$", "", value)  # noqa: PLW2901
```

Comment:
It would be a bit nicer to do this without any more `re` if you can. Something like this:
```
code = code[import_match.end():] # to remove imports
code.rpartition("\n\n") # something like this to remove the last line
```